constraint-solver/src/solver.rs

```diff
@@ -88,83 +88,42 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Solver<T, V>
 
     fn apply_effect(&mut self, effect: Effect<T, V>) -> bool {
         match effect {
-            Effect::Assignment(v, expr) => self.apply_assignment(v, expr),
+            Effect::Assignment(v, expr) => self.apply_assignment(&v, &expr),
             Effect::RangeConstraint(v, range_constraint) => {
-                self.apply_range_constraint_update(v, range_constraint)
+                self.apply_range_constraint_update(&v, range_constraint)
             }
             Effect::BitDecomposition(..) => unreachable!(),
             Effect::Assertion(..) => unreachable!(),
             Effect::ConditionalAssignment { .. } => todo!(),
         }
     }
 
-    fn apply_assignment(&mut self, variable: V, expr: SymbolicExpression<T, V>) -> bool {
-        let SymbolicExpression::Concrete(value) = expr else {
-            panic!("Unexpected non-concrete assignment: {variable} = {expr}");
-        };
-
-        let entry = self.variable_states.entry(variable.clone()).or_default();
-        let variable_update = match &entry {
-            // We already know the expression
-            VariableState::Known(existing) => {
-                assert_eq!(
-                    *existing, value,
-                    "Inconsistent assignment for {variable}: {existing} != {value}"
-                );
-                return false;
-            }
-
-            // An unknown variable became known.
-            VariableState::Unknown(..) => {
-                log::trace!("{variable} = {value}");
-
-                *entry = VariableState::Known(value);
-
-                // The borrow checker won't let us call `update_constraints` here, so we
-                // just return the update.
-                Some(VariableUpdate {
-                    variable,
-                    known: true,
-                    range_constraint: RangeConstraint::from_value(value),
-                })
-            }
-        };
-
-        if let Some(variable_update) = variable_update {
-            self.update_constraints(&variable_update);
-            true
-        } else {
-            false
-        }
+    fn apply_assignment(&mut self, variable: &V, expr: &SymbolicExpression<T, V>) -> bool {
+        self.apply_range_constraint_update(variable, expr.range_constraint())
```

Comment:
With #2667 , we can actually apply the substitution if it is not a compile-time constant.